internal/app/secretless/tcp_proxy_service.go

```diff
@@ -0,0 +1,147 @@
+			backendCredentials[credential][byteIndex] = 0
+		}
+	}
+}
+
+func duplexStream(source io.ReadWriter, destination io.ReadWriter) <- chan error {
+	errChan := make(chan error, 1)
+	go func() {
+		errChan <- stream(source, destination)
+	}()
+	go func() {
+		errChan <- stream(destination, source)
+	}()
+	return errChan
+}
+
+func stream(source io.Reader, destination io.Writer) error {
+	for {
+		if _, err := io.Copy(destination, source); err != nil {
+			return err
+		}
+		return nil
+	}
+}
+
+type tcpProxyService struct {
+	connector          tcp.Connector
+	resolveCredentials credentialsResolver
+	// TODO: use logger to report errors
+	// logger 			   log.Logger
+	listener           net.Listener
+	done               bool
+	error              chan error
+}
+
+// NewTCPProxyService constructs a new instance of a TCP ProxyService. The
+// constructor takes a TCP Connector, CredentialResolver and Listener.
+// A TCP ProxyService is able to Connect with Credentials then subsequently stream
+// bytes between client and target service
+func NewTCPProxyService(
+	connector tcp.Connector,
+	resolveCredentials credentialsResolver,
+	listener net.Listener,
+	) (internal.ProxyService, error) {
+	errors := validation.Errors{}
+
+	if connector == nil {
+		errors["connector"] = fmt.Errorf("connector cannot be nil")
+	}
+	if resolveCredentials == nil {
+		errors["resolveCredentials"] = fmt.Errorf("resolveCredentials cannot be nil")
+	}
+	if listener == nil {
+		errors["listener"] = fmt.Errorf("listener cannot be nil")
+	}
+
+	if err := errors.Filter(); err != nil {
+		return nil, err
+	}
+
+	return &tcpProxyService{
+		connector:          connector,
+		resolveCredentials: resolveCredentials,
+		listener:           listener,
+		done:               false,
+		error:              make(chan error),
+	}, nil
+}
+
+type credentialsResolver func() (map[string][]byte, error)
+
+func (ps *tcpProxyService) handleConnection(clientConn net.Conn) error {
+	var targetConn net.Conn
+	defer func() {
+		if clientConn != nil {
+			_ = clientConn.Close()
+		}
+		if clientConn != nil {
+			_ = targetConn.Close()
+		}
+	}()
+
+	backendCredentials, err := ps.resolveCredentials()
+	defer zeroizeCredentials(backendCredentials)
+	if err != nil {
+		return err
+	}
+
+	targetConn, err = ps.connector(clientConn, backendCredentials)
+	if err != nil {
+		return err
+	}
+
+	return <- duplexStream(clientConn, targetConn)
+}
+
+func (ps *tcpProxyService) OpError() chan error {
+	return ps.error
+}
+
+// Start initiates the net.Listener to listen for incoming connections
+func (ps *tcpProxyService) Start() error {
+	go func() { // n go routines for n tcp proxyservices
+		// TODO: add mutex to done
+		for !ps.done {
+			// TODO: can accepts happen in parallel ?
+			conn, err := ps.listener.Accept()
+			if err != nil {
+				ps.error <- err
+				return
+			}
+			go func() {
+				err := ps.handleConnection(conn)
+				ps.error <- err
+			}()
+		}
+	}()
+
+	return nil
```

Comment:
We might consider removing error return or return an error on socket creation.

Reply:
Not sure I follow. This method exists because this implementation needs to adhere to the ProxyService interface